### PR TITLE
[Lim Daekoon] Duke Increments

### PR DESCRIPTION
**Pull request by daekoon**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,92 @@
...
+            this.parser = new Parser("list");
+        } catch (Exception e) {
+            Ui.printBlock("Unexpected Error");
+        }
+
+    }
+
+    public void run() {
+
+        while (!parser.getType().equals("bye")) {
```
Comment (Icesiolz) [inline src/main/java/Duke.java]: I think this switch case is repeated in your Parser class, you might want to run the logic using the Parser class instead. (By creating the Command classes and passing the tasklist into the Commands' methods. 

Eg. 
In the parser:
public Command parse(String input) {
    switch (input):
    ...
    case "bye":
        return new ByeCommand;
    }
    ...
}

In command class:
public void execute(TaskList tasks, Ui ui, Storage storage) {
    ...
}

In duke:
Command c = parser.parse(input);
c.execute(tasks, ui, storage);

Comment (daekoon) [reply src/main/java/Duke.java]: Thanks for the input. I'll begin refactoring once I begin creating command classes.

**Review by wallacelim — APPROVED** [review]
Great code! I like how specific and modular things are throughout your code, and how OOP principles are well-adhered to (e.g. specific exceptions extending DukeException). Naming conventions are also apparently followed, and code style is consistent.

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,174 @@
...
+
+    /**
+     * Runs the Duke program.
+     */
+    public static void run() {
+
+        try {
+            parser = Ui.readCommand();
+
+            switch (parser.getType()) {
```
Comment (limerencee) [inline src/main/java/Duke.java:51]: Can consider using `Enumeration`.

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,174 @@
...
+            e.printMessage();
+        }
+    }
+
+    /**
+     * Default static main method of this project.
+     */
+
+    @Override
+    public void start(Stage stage) {
```
Comment (limerencee) [inline src/main/java/Duke.java:92]: Consider using Java FXML for cleaner-looking code!

File: src/main/java/Parser.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    private String description;
+    private String time;
+    private Integer index;
+    private String searchTerm;
+    private String sortBy;
+    private String sortOrder;
+
+    /**
+     * Constructs a Parser object based on input string.
+     * @param input The command String to be parsed.
```
Comment (limerencee) [inline src/main/java/Parser.java:32]: JavaDocs: There should be a new line after the method description and before the parameters.

File: src/main/java/Parser.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        this.index = null;
+        this.searchTerm = null;
+        this.sortBy = null;
+        this.sortOrder = null;
+
+        String[] inputs = input.split(" ", 2);
+        String[] params;
+
+        this.type = inputs[0];
+        switch (type) {
```
Comment (limerencee) [inline src/main/java/Parser.java:48]: Consider abstracting the cases to conform to SLAP.

File: src/main/java/Storage.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    private BufferedReader bufferedReader;
+    private BufferedWriter bufferedWriter;
+    private SimpleDateFormat dateFormatter = new SimpleDateFormat("dd/MM/yyyy HHmm");
+
+    /**
+     * Constructs the Storage object with the directory of save file.
+     * @param dir Directory of the save file.
+     */
+    public Storage(String dir) {
+        assert dir == null;
```
Comment (limerencee) [inline src/main/java/Storage.java:36]: Shouldn't it be `assert != null` or `assert dir.equals("")` to check if it's empty?

File: src/main/java/Storage.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        try {
+            nextLine = bufferedReader.readLine();
+        } catch (IOException e) {
+            nextLine = null;
+        }
+        ArrayList<Task> taskList = new ArrayList<Task>();
+
+        // Savefile layout
+        // Event type|done or not done(1 or 0)|Description|Date(if applicable)
+        while (nextLine != null) {
```
Comment (limerencee) [inline src/main/java/Storage.java:63]: Consider SLAP to abstract.

File: src/main/java/TaskList.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        String regex = String.format("(.*)%s(.*)", parser.getSearchTerm());
+        for (int z = 0; z < taskList.size(); z++) {
+            currentTask = taskList.get(z);
+            if (currentTask.getTaskName().matches(regex)) {
+                matches.add(currentTask);
+            }
+        }
+
+        if (matches.size() < 1) {
+            Ui.printBlock("There were no matching tasks in your list.");
```
Comment (limerencee) [inline src/main/java/TaskList.java:182]: Consider declaring `final` variables for magic literals.

File: src/main/java/task/Task.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+    // Method used to get the name of the task
+    public String getTaskName() {
+        return this.taskName;
+    }
+
+    public abstract Date getDate();
+
+    private String getStatusIcon() {
+        return (isDone ? "\u2713" : "\u2718"); //return tick or X symbols
```
Comment (limerencee) [inline src/main/java/task/Task.java:37]: Checkstyle: Avoid using escaped unicodes.

**Review by limerencee — COMMENTED** [review]
This is done during Week 5 tutorial session. Good job overall, clean OO design!

**Pull request closed without merging**